### PR TITLE
Change BinaryHeap::append rebuild heuristic

### LINKED ISSUE #77433: Suboptimal performance of BinaryHeap::append
The [current implementation of `BinaryHeap::append`](https://github.com/rust-lang/rust/blob/8fe73e80d762bc575040239fc05fb1c612873554/library/alloc/src/collections/binary_heap.rs#L609) uses a heuristic based on the worst-case number of comparisons to determine between two strategies:

1. Calling `self.extend(other.drain())`, equivalent to pushing all elements of the other heap.
2. Appending the vector of the other heap and rebuilding the heap from scratch.

I've done some benchmarking, and it turns out that method 1 (based on `extend`) is always faster for two heaps based on randomly shuffled data &ndash; on the computers I've tested it on, anyway. I've included images of [the benchmarks](https://github.com/hanmertens/dary_heap/blob/3fe25525773f42d53174599d54490f001e4929af/benches/dary_heap.rs) below: first the current `append` strategy, then the `extend` strategy (method 1). In the benchmarks, two heaps are merged, one containing the number of elements on the *x*-axis, the other containing 100,000 minus that number (for the rightmost data points both heaps are equal in size). The red line (sometimes hiding behind the green one) is in both images that corresponding to `std::collections::BinaryHeap`, the other lines are not relevant to this issue (they aren't standard library types).

append
extend

From the jump in performance in the first graph, you can clearly see when the switch between the two approaches happens. Graphs are similar for smaller heaps, I didn't test larger ones. It's possible method 2 is faster under other circumstances (maybe if one of the heaps contains mostly very small/large elements?), especially if both heaps are (almost) equal in size. However, I think the benchmark is more close to the "average" real-world use case than such a case would be (I'd be happy to be proven wrong about this, though).

<details><summary>Simplification of the benchmark that only runs the parts relevant for this issue</summary>

##### `Cargo.toml`
```toml
[package]
name = "binary_heap_bench"
version = "0.1.0"
edition = "2018"

[dependencies]
criterion = "0.3"
rand = "0.7"

[[bench]]
name = "benchmark"
harness = false
```

##### `benches/benchmark.rs`
```rust
use criterion::*;
use rand::{seq::SliceRandom, thread_rng};
use std::collections::BinaryHeap;
use std::iter::FromIterator;
use std::mem;

/// Data type we want to use
type T = u32;

/// Produce shuffled vector containing values 0..n
fn random_data(n: T) -> Vec<T> {
    let mut data = Vec::from_iter(0..n);
    data.shuffle(&mut thread_rng());
    data
}

fn two_heaps<H: From<Vec<T>>>(mut vec1: Vec<T>, i: usize) -> (H, H) {
    let vec2 = vec1.split_off(i);
    (vec1.into(), vec2.into())
}

fn std_heap_append((mut heap1, mut heap2): (BinaryHeap<T>, BinaryHeap<T>)) -> BinaryHeap<T> {
    heap1.append(&mut heap2);
    heap1
}

fn std_heap_extend((mut heap1, mut heap2): (BinaryHeap<T>, BinaryHeap<T>)) -> BinaryHeap<T> {
    if heap1.len() < heap2.len() {
        mem::swap(&mut heap1, &mut heap2);
    }
    heap1.extend(heap2.drain());
    heap1
}

fn benchmark(c: &mut Criterion) {
    let mut group = c.benchmark_group("append/extend");
    let base = 100000;
    let step = 2500;
    for i in (step..=base as usize / 2).step_by(step) {
        let size = BatchSize::SmallInput;
        group.bench_function(BenchmarkId::new("append", i), |b| {
            b.iter_batched(|| two_heaps(random_data(base), i), std_heap_append, size)
        });
        group.bench_function(BenchmarkId::new("extend", i), |b| {
            b.iter_batched(|| two_heaps(random_data(base), i), std_heap_extend, size)
        });
    }
}

criterion_group!(benches, benchmark);
criterion_main!(benches);
```
</details>

### PR DESCRIPTION
**Pull request by hanmertens**
This is faster, see #77433.

Fixes #77433

### CODE REVIEW

**Comment by jyn514** [comment]
r? @scottmcm 

https://github.com/rust-lang/rust/issues/77433#issuecomment-703158287

**Comment by scottmcm** [comment]
Given your comment in https://github.com/rust-lang/rust/issues/77433#issuecomment-723674952 that

> I guess it doesn't make much sense to merge #77435 as-is because of that.

I've marked this as waiting-on author for now.

Once you've made any changes you think are necessary, please run the following to mark it as needing review again:
```
@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author
```

**Comment by JohnCSimon** [comment]
Ping from triage - 
@hanmertens Can you please post your status on this PR?

**Comment by hanmertens** [comment]
As discussed in the comments of #77433 the heuristic should be changed (and not removed, as this PR currently does). I've done some benchmarks and I'll update the issue with that later this week. I hope that will shed some light on what the heuristic should be, then the PR can be updated.

**Comment by crlf0710** [comment]
@hanmertens Ping from triage, any updates on this?

File: library/alloc/src/collections/binary_heap.rs [hunk 1]
```diff
@@ -630,10 +630,16 @@ impl<T: Ord> BinaryHeap<T> {
         // and about 2 * (len1 + len2) comparisons in the worst case
         // while `extend` takes O(len2 * log(len1)) operations
         // and about 1 * len2 * log_2(len1) comparisons in the worst case,
-        // assuming len1 >= len2.
+        // assuming len1 >= len2. For larger heaps, the crossover point
+        // no longer follows this reasoning and was determined empirically.
         #[inline]
```
Comment (jyn514) [inline library/alloc/src/collections/binary_heap.rs:635]: Should this still be marked `#[inline]` now that it's longer?

**Comment by hanmertens** [comment]
I haven't received negative feedback on https://github.com/rust-lang/rust/issues/77433#issuecomment-744081846, so I just pushed the heuristic I proposed there. TLDR; it's not optimal but better than the current one.

File: library/alloc/src/collections/binary_heap.rs (hunk 1 above)
Comment (hanmertens) [reply library/alloc/src/collections/binary_heap.rs:635]: It's only used once, so that should still be fine I think.

**Comment by scottmcm** [comment]
Thanks!  Seems reasonable to me, based on the other issue.

@bors r+ rollup=maybe

(I don't think the compiler uses binary heaps enough for this to affect its perf, so doesn't need to be `=never`.)

**Comment by jyn514** [comment]
> (I don't think the compiler uses binary heaps enough for this to affect its perf, so doesn't need to be `=never`.)

```
> rg BinaryHeap compiler/ | wc -l
0
```
You are correct :laughing:

**Comment by jyn514** [comment]
In fact the *only* place `BinaryHeap` is used in-tree is in tests for clippy and miri, and a few related clippy lints.

**Pull request merged**